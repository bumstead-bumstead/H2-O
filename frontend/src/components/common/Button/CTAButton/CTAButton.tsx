import type { ButtonHTMLAttributes, MouseEventHandler } from 'react';
import { css } from '@emotion/react';
import styled from '@emotion/styled';

type ButtonSize = 'small' | 'medium' | 'large';

interface Props extends ButtonHTMLAttributes<HTMLButtonElement> {
  isFull?: boolean;
  size?: ButtonSize;
}

function CTAButton({ children, size = 'medium', onClick, ...restProps }: Props) {
  const handleClick: MouseEventHandler<HTMLButtonElement> = (e) => {
    onClick?.(e);
    const target = e.currentTarget;
    target.style.pointerEvents = 'none';

    setTimeout(() => {
      target.style.pointerEvents = 'auto';
    }, 500);
  };

  return (
    <StyledButton size={size} onClick={handleClick} {...restProps}>
      {children}
    </StyledButton>
  );
}

export default CTAButton;

const StyledButton = styled.button<Pick<Props, 'size' | 'isFull'>>`
  ${({ theme }) => theme.flex.flexCenterRow}
  ${({ theme }) => theme.typography.TextKRMedium16}
  background-color: ${({ theme }) => theme.colors.primary700};
  color: ${({ theme }) => theme.colors.white};
  border-radius: 4px;

  ${({ theme, size }) => {
    switch (size) {
      case 'small':
        return css`
          width: 220px;
          height: 36px;
        `;
      case 'medium':
        return css`
          ${theme.typography.HeadKRMedium16}
          width: 340px;
          height: 44px;
        `;
      case 'large':
        return css`
          width: 850px;
          height: 52px;
        `;
    }
  }}

  ${({ isFull }) =>
    isFull &&
    css`
      width: 100%;
    `}

  &:hover {
    background-color: ${({ theme }) => theme.colors.primary500};
  }

  &:active {
    background-color: ${({ theme }) => theme.colors.primary800};
  }

  &:disabled {
    background-color: ${({ theme }) => theme.colors.gray300};
    cursor: not-allowed;
  }
`;
