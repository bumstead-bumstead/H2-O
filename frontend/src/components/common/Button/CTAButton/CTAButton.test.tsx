import CTAButton from './CTAButton';
import { render, screen, userEvent, waitFor } from '@/tests/test-util';

describe('CTAButton 컴포넌트 테스트', () => {
  it('버튼이 정상적으로 렌더링된다.', () => {
    render(<CTAButton>버튼</CTAButton>);

    expect(screen.getByRole('button')).toBeInTheDocument();
  });

  it('버튼의 onClick 핸들러가 클릭시 정상적으로 호출된다.', async () => {
    const mockOnClick = vi.fn();
    render(<CTAButton onClick={mockOnClick}>버튼</CTAButton>);

    userEvent.click(screen.getByRole('button'));
    await waitFor(() => expect(mockOnClick).toBeCalledTimes(1));
  });

  it('버튼을 다중 클릭할 시 핸들러가 한번만 호출된다.', async () => {
    const mockOnClick = vi.fn();
    render(<CTAButton onClick={mockOnClick}>버튼</CTAButton>);

    userEvent.tripleClick(screen.getByRole('button'));
    await waitFor(() => expect(mockOnClick).toBeCalledTimes(1));
  });

  it('버튼이 disabled인 경우 클릭할 수 없다.', async () => {
    const mockOnClick = vi.fn();
    render(
      <CTAButton disabled onClick={mockOnClick}>
        버튼
      </CTAButton>,
    );

    const button = screen.getByRole('button');
    expect(button).toBeDisabled();

    userEvent.click(button);
    await waitFor(() => expect(mockOnClick).not.toBeCalled());
  });

  it('버튼의 isFull 속성이 true일 경우 width가 100%이다.', () => {
    render(<CTAButton isFull>버튼</CTAButton>);

    expect(screen.getByRole('button')).toHaveStyle('width: 100%');
  });

  describe('CTAButton 스냅샷 테스트', () => {
    it('size 속성이 small일 때 스냅샷이 일치한다.', () => {
      render(<CTAButton size='small'>버튼</CTAButton>);

      expect(screen.getByRole('button')).toMatchSnapshot();
    });

    it('size 속성이 medium일 때 스냅샷이 일치한다.', () => {
      render(<CTAButton size='medium'>버튼</CTAButton>);

      expect(screen.getByRole('button')).toMatchSnapshot();
    });

    it('size 속성이 large일 때 스냅샷이 일치한다.', () => {
      render(<CTAButton size='large'>버튼</CTAButton>);

      expect(screen.getByRole('button')).toMatchSnapshot();
    });
  });
});
