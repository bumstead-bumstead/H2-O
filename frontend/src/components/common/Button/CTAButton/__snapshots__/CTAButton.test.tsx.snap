// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CTAButton 컴포넌트 테스트 > CTAButton 스냅샷 테스트 > size 속성이 large일 때 스냅샷이 일치한다. 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-family: Hyundai Sans Text KR,sans-serif;
  font-size: 16px;
  font-weight: 500;
  line-height: 18px;
  letter-spacing: -0.03em;
  background-color: #00224A;
  color: #FFFFFF;
  border-radius: 4px;
  width: 850px;
  height: 52px;
}

.emotion-0:hover {
  background-color: #00397B;
}

.emotion-0:active {
  background-color: #001731;
}

.emotion-0:disabled {
  background-color: #C8CACC;
  cursor: not-allowed;
}

<button
  class="emotion-0"
>
  버튼
</button>
`;

exports[`CTAButton 컴포넌트 테스트 > CTAButton 스냅샷 테스트 > size 속성이 medium일 때 스냅샷이 일치한다. 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-family: Hyundai Sans Text KR,sans-serif;
  font-size: 16px;
  font-weight: 500;
  line-height: 18px;
  letter-spacing: -0.03em;
  background-color: #00224A;
  color: #FFFFFF;
  border-radius: 4px;
  font-family: Hyundai Sans Head KR,sans-serif;
  font-size: 16px;
  font-weight: 500;
  line-height: 24px;
  letter-spacing: -0.03em;
  width: 340px;
  height: 44px;
}

.emotion-0:hover {
  background-color: #00397B;
}

.emotion-0:active {
  background-color: #001731;
}

.emotion-0:disabled {
  background-color: #C8CACC;
  cursor: not-allowed;
}

<button
  class="emotion-0"
>
  버튼
</button>
`;

exports[`CTAButton 컴포넌트 테스트 > CTAButton 스냅샷 테스트 > size 속성이 small일 때 스냅샷이 일치한다. 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-family: Hyundai Sans Text KR,sans-serif;
  font-size: 16px;
  font-weight: 500;
  line-height: 18px;
  letter-spacing: -0.03em;
  background-color: #00224A;
  color: #FFFFFF;
  border-radius: 4px;
  width: 220px;
  height: 36px;
}

.emotion-0:hover {
  background-color: #00397B;
}

.emotion-0:active {
  background-color: #001731;
}

.emotion-0:disabled {
  background-color: #C8CACC;
  cursor: not-allowed;
}

<button
  class="emotion-0"
>
  버튼
</button>
`;
